src/EditorFeatures/Core/Diagnostics/AbstractPushOrPullDiagnosticsTaggerProvider.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Editor.Shared.Tagging;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Editor.Tagging;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Workspaces;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Tagging;
+
+namespace Microsoft.CodeAnalysis.Diagnostics;
+
+internal static class DiagnosticTaggingOptions
+{
+    public static readonly Option2<bool> PullDiagnosticTagging = new(
+        "DiagnosticTaggingOptions", "PullDiagnosticTagging", defaultValue: true,
+        new FeatureFlagStorageLocation("Roslyn.PullDiagnosticTagging"));
+}
+
+/// <summary>
+/// Base type of all diagnostic taggers (classification, squiggles, suggestions, inline-diags).  Subclasses can control
+/// things by overriding functionality in this type.  Internally, this will switch to either a pull or push cased
+/// approach at instantiation time depending on our internal feature flag.
+/// </summary>
+internal abstract partial class AbstractPushOrPullDiagnosticsTaggerProvider<TTag> : ITaggerProvider
+    where TTag : ITag
+{
+    private readonly ITaggerProvider _underlyingTaggerProvider;
+
+    protected readonly IGlobalOptionService GlobalOptions;
+
+    protected AbstractPushOrPullDiagnosticsTaggerProvider(
+        IThreadingContext threadingContext,
+        IDiagnosticService diagnosticService,
+        IDiagnosticAnalyzerService analyzerService,
+        IGlobalOptionService globalOptions,
+        ITextBufferVisibilityTracker? visibilityTracker,
+        IAsynchronousOperationListener listener)
+    {
+        GlobalOptions = globalOptions;
+
+        if (globalOptions.GetOption(DiagnosticTaggingOptions.PullDiagnosticTagging))
+        {
+            _underlyingTaggerProvider = new PullDiagnosticsTaggerProvider(
+                this, threadingContext, diagnosticService, analyzerService, globalOptions, visibilityTracker, listener);
+        }
+        else
+        {
+            _underlyingTaggerProvider = new PushDiagnosticsTaggerProvider(
+                this, threadingContext, diagnosticService, globalOptions, visibilityTracker, listener);
+        }
```

Comment:
this is where we make the high-level decision *immediately* about if this feature will be using push or pull.  Effectively, this tagger-provider just wraps another tagger provider just to give us that switch point.